### ISSUE TITLE
How to use a constant Matrix as Input Data?

### ISSUE THREAD

**Issue by 10132510143cai** [dmlc/mxnet#5754]
Hi, I'm trying to add a constant Matrix into my neual network. This is part of my network design

`   xipart_net = mx.sym.Variable(name='dataxi')
    xipart_net = mx.sym.FullyConnected(data=xipart_net, weight=fc1_weight, bias=fc1_bias, num_hidden=64, name="fc1")
    xipart_net = mx.sym.Activation(xipart_net, name='relu1', act_type="relu")
    xipart_net = mx.sym.Dropout(data=xipart_net, p=0.2)
    xipart_net = mx.sym.FullyConnected(data=xipart_net, weight=fc2_weight, bias=fc2_bias, num_hidden=k, name="fc2")  # dimension batch_size*k

    m = mx.sym.Variable(name='M', shape=(10, 10))  # M k*k

    xipart_net = mx.symbol.dot(lhs=xipart_net, rhs=m)  # dimension batchsize * k`
    

But I cannot Input the M with the shape (10, 10). I set the shape of M (10,10),but it told me

infer_shape error. Arguments:
  M: (64, 10)
  dataxi: (64, 784)
  dataxj: (64, 784)
  isinM: (64,)

64 is my batchsize. The Next is myDataIter, and i hope how to use a constant Matrix as input Data? Thank u.

`class MyDataIter(mx.io.DataIter):
    def __init__(self, z, label, batch_size):
        super(MyDataIter, self).__init__()
        self.data = z
        self.label = label
        self.batch_size = batch_size
        self.provide_data = [('dataxi', (batch_size,  784)), ('dataxj', (batch_size,  784)), ('isinM', (batch_size, )),('M',(10, 10))]
        self.provide_label = []`

**Comment by bhokaal2k** [comment]
During shape inference step MxNet checks for consistency in the input shapes w.r.t. batch size and requires that the outermost dimension for all the inputs must be equal to the batch_size and hence the error, because M's shape is (10,10). If M is a constant why not input it as a constant Variable initialized with the required value? Or else, simply make it (batch_size, 10, 10) and while using it for the dot product first take out one of the channels using slice_channel() and then use that symbol in the mx.symbol.dot().

**Comment by 10132510143cai** [comment]
At first, I also want to make it  (batch_size, 10, 10) but I don't know how to slice it to (10, 10) and put it in the symbol.

Now, I set M as the param and set its learning rate=0,
`m = mx.sym.Variable(name='M', shape=(10, 10), attr={'lr_mult': '0'})`
and set M at the feedforward
`model = mx.model.FeedForward(
            symbol=net,  # network structure
            num_epoch=num_epoch,  # number of data passes for training
            learning_rate=learning_rate,  # learning rate of SGD
            initializer=mx.init.Xavier(factor_type="in", magnitude=2.34),
            arg_params={'M': M}
        )`

It seems work. Thank u for your comment.
